backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchSearchPredicate.java

```diff
@@ -22,13 +22,13 @@ public abstract class ElasticsearchSearchPredicate implements SearchPredicate {
 
 	public abstract Set<String> indexNames();
 
-	public abstract void checkNestableWithin(String expectedParentNestedPath);
-
-	public final JsonObject toJsonQuery(PredicateRequestContext context) {
-		return buildJsonQuery( context.predicateContext( this ) );
+	public final void checkNestableWithin(PredicateNestingContext context) {
+		doCheckNestableWithin( context.wrap( this ) );
```

Comment:
Consider this case:
```java
// all good we do not call any `checkNestableWithin` as this knn is at the very top level:
SearchPredicate knn = scope.predicate().knn( 15 ).field( "location" ).matching( 50.0f, 50.0f ).toPredicate();
// here we add it in a should clause so we know it can accept the knn 
// and we create a `PredicateNestingContext.acceptsKnn()` context:
SearchPredicate boolKnnInShould = scope.predicate().bool().should( knn )
		.should( scope.predicate().match().field( "parking" ).matching( Boolean.TRUE ) ).toPredicate();
// now here ...  
// the first should will create a `PredicateNestingContext.acceptsKnn()` context. 
// i.e it will receive a boolean predicate on which we are calling `imaBoolPredicateWithKnnShouldClause.checkNestableWithin( PredicateNestingContext.acceptsKnn() );`
// so this first call is ok, but then any inner predicates of this `imaBoolPredicateWithKnnShouldClause` 
// should not allow knn preidcates, right ? soo when this is executed:
// 	@Override
// 	public void doCheckNestableWithin(PredicateNestingContext context) {
// 		checkAcceptableWithin( context, mustClauses );
//             // context here now should not allow accepting knn predicates, so it should somehow be updated... 
// 		checkAcceptableWithin( context, shouldClauses );
// 		checkAcceptableWithin( context, filterClauses );
// 		checkAcceptableWithin( context, mustNotClauses );
// 	}
scope.predicate().bool().should( boolKnnInShould );
```

we probably could try replacing `doCheckNestableWithin` in a bool predicate with something like
```java
@Override
public void doCheckNestableWithin(PredicateNestingContext context) {
	PredicateNestingContext updatedContext = PredicateNestingContext.doesNotAcceptKnn( context.getNestedPath() );
	checkAcceptableWithin( updatedContext, mustClauses );
	checkAcceptableWithin( updatedContext, shouldClauses );
	checkAcceptableWithin( updatedContext, filterClauses );
	checkAcceptableWithin( updatedContext, mustNotClauses );
}
```
but then that wrap is doing it for us already

Reply:
you can check this test to see how it behaves 
https://github.com/hibernate/hibernate-search/blob/cbda56b8ab9a36637d1ae8210e2d11228a3bbf5d/integrationtest/backend/elasticsearch/src/test/java/org/hibernate/search/integrationtest/backend/elasticsearch/search/ElasticsearchKnnPredicateSpecificsIT.java#L182-L209

Reply:
I'd actually have expected your specific example (a `knn` within a `should` within a `should`) to work without throwing an exception... After all, `X OR (Y OR Knn)` is the same as `Knn OR (X OR Y)`.

Though I see the documentation does mention something about "top-level"... any reason to that? Score when `must` clauses are added to the mix, maybe?

I mean if it's possible to support knn predicates in nested should clauses *and* it simplifies implementation, I'd definitely be in favor.
And personally even if we don't want to support it, I'd favor a solution that doesn't involve double dispatch if it's not much more complicated.